### ISSUE TITLE
Database library improvements

### ISSUE THREAD

**Issue by euantorano** [mybb/mybb#4206]
Once we get #4193 merged, I have a few proposed alterations to make to the DB libraries:

- [ ] Add some new fetch methods for fetching results
    - [ ] `fetchAllArray` - fetch all results as an array of arrays
    - [ ] `fetchAllField` - fetch a single column from all resulting rows
    - [ ] `fetchObject` - fetch a result row as a class of the given type - this will be useful in the future when using typed properties to eliminate boilerplate
    - [ ] `fetchAllObject` - fetch all results as an array of objects
- [ ] Add some new methods to make use of prepared statements - with fallbacks for drivers that don't support prepared statements - such as:
   - [ ] `insertQueryPrepared` - the same as `insert_query`, but the `$array` parameter is bound rather than escaped by the caller and interpolated.
   - [ ] `updateQueryPrepared` - similar to `update_query`, but the `$array`, `$where` and `$limit` parameters are bound rather than escaped by the caller and interpolated.
   - [ ] `deleteQueryPrepared` - similar to `delete_query`, but the `$where` and `$limit` parameters are bound rather than escaped by the caller and interpolated.
   - [ ] `simpleSelectPrepared` - similar to `simple_select`, but the `$conditions` and `$options` parameters are bound rather than escaped by the caller and interpolated.

I'd like to be able to write DB code like this and know that I don't need to worry about things like `$db->escape_string`:

```php
$query = $db->insertQueryPrepared('datacache', array('title' => 'version', 'cache' => '1.8.26'));
```

```php
$query = $db->simplSelectPrepared('users', '*', array('postnum' => 1), array('limit' => 1));

$usersWithOnePost = $db->fetchAllObject($query, User::class);
```

Ideally I'd like to get these improvements into 1.8, as they wouldn't break backwards compatibility in the core, and we could drive plugin developers towards some more modern DB code by doing so.

This would be aimed at 1.8.26, rather than 1.8.25 (so the person after next, not next).

Note that this would also be beneficial for the merge system if we could expose the ability to prepare a statement and re-use the prepared statement with different bound parameters.

**Comment by dragonexpert** [comment]
Not sure if you are aware that fetch_array has a second parameter which determines what result type to return, I will have to look into if Object is a valid type.  The value is the constant name for the class such as MYSQLI_ASSOC.  I will at least work on the fetch methods though since those shouldn't be too time consuming.  I'll have to refresh myself on how to do prepared statements because it has been a while since I used them last.

**Comment by euantorano** [comment]
Yeah, I was aware of fetch_array()’s second parameter but the name is a bit misleading if we use it to fetch objects. 

If you need any help or anything just let me know :)

**Comment by dragonexpert** [comment]
I have the various fetch functions completed.  I have tested mysqli because that is what my environment is.  [My Output](http://teamdimensional.net/testforums/test.php).  I do not have an environment set up to test the other database classes, but on my fork, I did the changes to the various db files.

**Comment by dragonexpert** [comment]
I'm thinking for the prepared statements, that there should be one generic prepare_statement function.  
function prepare_statement($sql, $params);

The sql should be generated by the explicit method.  $params would have to be an associative array of column_name => column_value.  This way we can recycle code since it can get lengthy really fast and it would reduce the chance of errors if all the binding is done in one place.

**Comment by euantorano** [comment]
@dragonexpert Looks good so far! 

RE binding: I think we need to have an option to simply prepare without parameters and bind parameters at a later stage. My thought process for this is for operations like bulk inserts - here, it is useful to prepare a statement once, then execute it multiple times with different data each time. Here's an example API:

```php
$query = $db->prepare('UPDATE mybb_users SET postnum = 0 WHERE uid = :uid');

foreach ([1,2,3,4,5,6,7,8,9,10] as $uid) {
    $db->executePrepared($query, ['uid' => $uid]);
}
```

In order to make this work with the `explain` (used for the debug page output), we might need to write a class to wrap the `PDOStatement` class. It may be worth looking at how other projects do this, like [`php-debugbar`](https://github.com/maximebf/php-debugbar/tree/master/src/DebugBar/DataCollector/PDO).

**Comment by dragonexpert** [comment]
Can I have someone test db_pgsql.php and db_sqlite.php?  The main one I'm not certain about is PGSQL because PHPStorm complains that I have too many arguments for pg_fetch_object, but php.net says it can take 4.

**Comment by effone** [comment]
Sorry for jumping in, but:
```
$query = $db->simplSelectPrepared('users', '*', array('postnum' => 1), array('limit' => 1));
$usersWithOnePost = $db->fetchAllObject($query, User::class);
```

Instead of the above; the following is more friendly to me:
```
$result = $db->table('users')->select('*')->where('postnum', '=', '1')->limit('1')->fetchAll();
```

**Comment by euantorano** [comment]
@effone I agree, but that's probably something for further down the line as it's a much bigger change in syntax. That syntax is similar to other query builders like [Atlas.Query](https://github.com/atlasphp/Atlas.Query) and is the direction I'd like to head in in the future.

**Comment by euantorano** [comment]
> Can I have someone test db_pgsql.php and db_sqlite.php? The main one I'm not certain about is PGSQL because PHPStorm complains that I have too many arguments for pg_fetch_object, but php.net says it can take 4.

I'll do some testing either today or tomorrow, no problem 👍
